### PR TITLE
feat(fcm): Added support for specifying the analytics label for notifications.

### PR DESCRIPTION
**Pull request by chemidy**
- add analytics_label
#580

### CODE REVIEW

**Comment by chemidy** [comment]
I have already done the feature, i'm just waiting for the merge of the analytics branch because i've based my branch on it :)

File: src/messaging/messaging-types.ts [hunk 1]
```diff
@@ -60,6 +61,10 @@ export interface Notification {
   body?: string;
 }
 
+export interface FcmOptions {
```
Comment (hiranya911) [inline src/messaging/messaging-types.ts]: I think there should be code in place to validate the analytics label in each of these structs. Similar to how we validate other fields during json marshaling.

File: test/unit/messaging/messaging.spec.ts [hunk 2]
```diff
@@ -2750,6 +2750,14 @@ describe('Messaging', () => {
           },
         },
       },
+      {
+        label: 'Generic fcmOptions message',
+        req: {
+          fcmOptions: {
+            analyticsLabel: 'test.analytics',
```
Comment (hiranya911) [inline test/unit/messaging/messaging.spec.ts]: I think these should be serialized as `analytics_label`. @chong-shao can you confirm if `analyticsLabel` is also acceptable?

File: src/index.d.ts [hunk 3]
```diff
@@ -3701,6 +3707,17 @@ declare namespace admin.messaging {
     channelId?: string;
   }
 
+  /**
+   * Represents Options for features provided by the FCM SDK for Android.
```
Comment (egilmorez) [inline src/index.d.ts]: Nit: I don't think this needs to be capitalized.

File: src/index.d.ts [hunk 4]
```diff
@@ -3701,6 +3707,17 @@ declare namespace admin.messaging {
...
+   */
+  interface AndroidFcmOptions {
+
+    /**
+     * The Label associated with the message's analytics data..
```
Comment (egilmorez) [inline src/index.d.ts]: Nit: I don't think this needs to be capitalized.

File: src/index.d.ts [hunk 5]
```diff
@@ -3816,6 +3838,28 @@ declare namespace admin.messaging {
     volume?: number;
   }
 
+  /**
+   * Represents Options for features provided by the FCM SDK for iOS..
```
Comment (egilmorez) [inline src/index.d.ts]: Nit: I don't think this needs to be capitalized.

Comment (egilmorez) [inline src/index.d.ts]: Please remove extra period.

File: src/index.d.ts [hunk 6]
```diff
@@ -3816,6 +3838,28 @@ declare namespace admin.messaging {
...
+   */
+  interface ApnsFcmOptions {
+
+    /**
+     * The Label associated with the message's analytics data..
```
Comment (egilmorez) [inline src/index.d.ts]: Nit: I don't think this needs to be capitalized.

File: src/index.d.ts [hunk 7]
```diff
@@ -3816,6 +3838,28 @@ declare namespace admin.messaging {
...
+     */
+    analyticsLabel?: string;
+  }
+
+  /**
+   * Represents Platform independent options for features provided by the FCM SDKs.
```
Comment (egilmorez) [inline src/index.d.ts]: Nit: I don't think this needs to be capitalized.

It's an open question whether this term should be hyphenated ("platform-independent") when used as a modifier, so I leave that to you :)

File: src/index.d.ts [hunk 8]
```diff
@@ -3816,6 +3838,28 @@ declare namespace admin.messaging {
...
+   */
+  interface FcmOptions {
+
+    /**
+     * The Label associated with the message's analytics data..
```
Comment (egilmorez) [inline src/index.d.ts]: Nit: I don't think this needs to be capitalized.

Comment (egilmorez) [inline src/index.d.ts]: Please remove extra period.

File: src/index.d.ts (hunk 6 above)
Comment (egilmorez) [inline src/index.d.ts]: Please remove extra period.

File: src/index.d.ts (hunk 4 above)
Comment (egilmorez) [inline src/index.d.ts]: Please remove extra period.

File: src/messaging/messaging-types.ts (hunk 1 above)
Comment (chemidy) [reply src/messaging/messaging-types.ts]: Ok done

File: test/unit/messaging/messaging.spec.ts [hunk 9]
```diff
@@ -2852,6 +2860,9 @@ describe('Messaging', () => {
               bodyLocArgs: ['arg1', 'arg2'],
               channelId: 'test.channel',
             },
+            fcmOptions: {
+              analyticsLabel: 'test.analytics',
```
Comment (hiranya911) [inline test/unit/messaging/messaging.spec.ts]: Can you also add some test cases to make sure the new validators work as intended?

Comment (chemidy) [reply test/unit/messaging/messaging.spec.ts]: OK done

File: test/unit/messaging/messaging.spec.ts (hunk 2 above)
Comment (chong-shao) [reply test/unit/messaging/messaging.spec.ts]: I tried "analyticsLabel" in the HTTP API. It is also acceptable.

**Comment by parndt** [comment]
Hey, thanks for doing this! :+1: Is there a release coming soon with this commit included?

**Pull request merged**